File: self_deadlock.py
import threading

lock1 = threading.Lock()

def critical_section():
    lock1.acquire()
    print("Lock aquired first time")
    
    """
    Do some work here

    """
    print("Thread is trying to acquire the lock again, but it has not released it yet")
    try:
        lock1.acquire()
        print("Lock aquired second time")
    finally:
        lock1.release()

    lock1.release()


thread = threading.Thread(target=critical_section)
thread.start()
thread.join()
print("Thread finished")

File: cpp_implemts/self_deadlock.cpp
// C++ equivalent of self_deadlock.py
#include <iostream>
#include <thread>
#include <mutex>

std::mutex lock1;

void critical_section() {
    lock1.lock();
    std::cout << "[Thread] Lock acquired first time" << std::endl;
    
    // Simulate work in the critical section
    std::this_thread::sleep_for(std::chrono::seconds(1));

    std::cout << "[Thread] Trying to acquire the lock again" << std::endl;
    
    // This will cause a deadlock as the same thread tries to acquire the lock again
    lock1.lock();
    std::cout << "[Thread] Lock acquired second time (this line will never be reached)" << std::endl;

    lock1.unlock();
    lock1.unlock();
}

int main() {
    std::thread t(critical_section);
    t.join();
    
    std::cout << "[Main] Program finished" << std::endl;
    return 0;
}
